extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/WindowOperatorQueryKit.java

```diff
@@ -247,55 +227,47 @@ public QueryDefinition makeQueryDefinition(
   /**
    *
    * @param originalQuery
-   * @return A list of list of operator factories, where each list represents the operator factories for a particular
-   * window stage.
+   * @return A list of {@link WindowStage}.
    */
-  private List<List<OperatorFactory>> getOperatorListFromQuery(WindowOperatorQuery originalQuery)
+  private List<WindowStage> getWindowStagesFromQuery(WindowOperatorQuery originalQuery)
   {
-    List<List<OperatorFactory>> operatorList = new ArrayList<>();
+    final List<WindowStage> stages = new ArrayList<>();
     final List<OperatorFactory> operators = originalQuery.getOperators();
-    List<OperatorFactory> currentStage = new ArrayList<>();
+    WindowStage currentStage = new WindowStage();
 
     for (int i = 0; i < operators.size(); i++) {
```

Comment:
To be able to move this logic into the WindowStage class, the WindowStage would have to store some previous state IIUC.

Without previous state, we cannot make the decision of whether a NaiveSortOperator or a NaivePartitioningOperator should go into the same stage or create a new stage. (I'm also trying to avoid making any functional changes in this PR, and keeping it purely a refactoring change)

For example, for queries like
```sql
select countryName, cityName,
count(*) over(partition by countryName order by cityName) as c1,
row_number() over(partition by countryName, cityName order by cityName) as c2
from wikipedia
where cityName in ('New York', 'Aarhus')
group by countryName, cityName;
```

MSQ gets the list of operator factories as:
```
NaiveSortOperatorFactory(d1)
NaivePartitioningOperatorFactory(d1)
WindowOperatorFactory
NaivePartitioningOperatorFactory(d1,d0)
WindowOperatorFactory
```

So we need the previous state to determine whether to add `NaivePartitioningOperatorFactory(d1,d0)` to the current stage or if we need to create a new stage.

Also, I think we'd get a better picture of how this would/should look as part of the GlueingPartitioningOperator PR, so I aimed to keep it simple for this PR. Hope that works!

Reply:
Does the following logic look good for a method `WindowStage#canAccept()`?

```java
public boolean canAccept(OperatorFactory operatorFactory)
    {
      if (operatorFactory instanceof NaiveSortOperatorFactory) {
        return false; // Start a new stage
      }

      if (operatorFactory instanceof WindowOperatorFactory) {
        return true; // Add to current stage
      }

      // If it's a PartitioningOperatorFactory, we can add it to current stage if NaiveSortOperatorFactory is present.
      return sortOperatorFactory != null;
    }
```

(PS: I'm trying to avoid making any functional changes in this PR, and keeping it purely a refactoring change)

Reply:
it could be a start - but its buggy as it will never accept a `NaiveSortOperatorFactory` ?

Reply:
Currently, we start a new stage immediately after every "consecutive chain of WindowOperatorFactory".

This effectively means that we always start a new stage in 2 occasions:
1. On encountering NaiveSortOperatorFactory
2. On encountering NaivePartitioningOperatorFactory, but not having encountered a NaiveSortOperatorFactory

The code snippet I posted above represents this logic. (and it works, all tests pass)

This logic would have to be completely re-worked with the GlueingPartitioningOperator PR changes (I don't know how that will look like), which is why I was hoping to get the other PR merged first.